articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+    switch (status) {
+      case 'pending':
+        return (
+          <p>Loading</p>
+        )
+      case 'resolved':
+        return (
+          <ProductInfo product={productData!} />
+        )
+      case 'rejected':
+        return (
+          <div>{error}</div>
+        )
+      default:
+        return (
+          <p>Type a product ID</p>
+        )
+    }
+  }
+
+  return (
+    <div>
+      <div>
+        <input type="text" ref={inputRef} />
+        <button onClick={handleSearch}>Search</button>
+      </div>
+      <div>
+        {renderContent()}
+      </div>
+    </div>
+  )
+}
+```
+
+![product-search-whole-flow.gif](/images/articles/react-error-handling/product-search-whole-flow.gif)
+
+
+It works on a way better organized :). 
+
+Getting to the point of this whole post what if the backend team makes some release on the API and some productReference's value is no longer being returned (if you are a backend developer and you are reading this, don't hesitate this is just for testing purpose you are the best of the best). This affects our new functionallity to show some product details based on the productReference. Let's take a look how this behaves, we will not actually change anyything but removing a productReference's value.
+
+```ts
+const products = {
+  '1': {
+    productName: 'xxxx',
+    productReference: '1-xxxx',
+    price: 0
+  },
+  '2': {
+    productName: 'xxxx',
+    price: 0
+  }
+} as unknown as Record<string, ProductData>
+```
+The casting on the last line is to not update any ts interface. If we review our frontend result, we get the following:
+
+![product-search-uncaught-product-reference-error.gif](/images/articles/react-error-handling/product-search-uncaught-product-reference-error.gif)
+
+May it be our error? Yes, it may. Must we create a ticket to the backend team? As soon as possible please. While the error gets solved let's review the error handling pattern installing the next [dependecy](https://www.npmjs.com/package/react-error-boundary).
+
+```cmd
+npm install react-error-boundary
+```
+
+After installing the package we need to import the `ErrorBoundary` component and wrap our `renderContent` function in it
+
+```tsx
+import { ErrorBoundary } from 'react-error-boundary'
+
+// rest of the code
+
+<ErrorBoundary fallback={<p>There was an error</p>}>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+![product-search-error-boundary-fallback.gif](/images/articles/react-error-handling/product-search-error-boundary-fallback.gif)
+
+But wait, the uncaught error alert is still being thwron. Is'not this the problem we will solve? Yeah, it is. We are at development mode (using the react's dev libraries) so react throws that error by default it can not be hidden what it did change is the result after closing that alert, our component is still there and we can keep interacting somehow (at production the error alert is not shown, the benefits of ErrorBoundary are way more appreciable).
```

Comment:
```suggestion
But wait, the uncaught error alert is still being thrown. Isn't this the problem we will solve? Yeah, it is. We are at development mode (using the react's dev libraries) so react throws that error by default it can not be hidden, what it did change is the result after closing that alert, our component is still there and we can keep interacting with it somehow (at production, the error alert is not shown, the benefits of ErrorBoundary are way more appreciable).
```